src/background/dnr.js

```diff
@@ -21,29 +24,87 @@ if (__PLATFORM__ === 'chromium' || __PLATFORM__ === 'safari') {
   // Ensure that DNR rulesets are equal to those from options.
   // eg. when web extension updates, the rulesets are reset
   // to the value from the manifest.
-  OptionsObserver.addListener(async function dnr(options) {
+  OptionsObserver.addListener(async function dnr(options, lastOptions) {
     const globalPause = getPausedDetails(options);
 
     const ids = ENGINES.map(({ name, key }) => {
       return !globalPause && options.terms && options[key] ? name : '';
     }).filter((id) => id && DNR_RESOURCES.includes(id));
 
+    const enabledRulesetIds =
+      (await chrome.declarativeNetRequest.getEnabledRulesets()) || [];
+
+    // Add regional network filters
+    if (ids.length && options.regionalFilters.enabled) {
+      ids.push(
+        ...options.regionalFilters.regions
+          .map((id) => `lang-${id}`)
+          .filter((id) => DNR_RESOURCES.includes(id)),
+      );
+    }
+
+    // Add fixes rules
     if (ids.length) {
-      ids.push('fixes');
-
-      // Add regional network filters
-      if (options.regionalFilters.enabled) {
-        ids.push(
-          ...options.regionalFilters.regions
-            .map((id) => `lang-${id}`)
-            .filter((id) => DNR_RESOURCES.includes(id)),
-        );
+      const fixesRulesIds = await getDynamicRulesIds(FIXES_ID_RANGE);
```

Comment:
this run on every options update?

It should definitely not run if we already have rules that are up to date. We should also have a schedule to update fixes every hour, same as we do with engines.

Reply:
There is no other way to know if dynamic rules are there. This code must run on disable/enable engines operation, which requires to be in options observer.

Reply:
I've meant the full operation. Specifically the download.

Reply:
you can keep a flag if there was an update - you will probably need one to prevent updating too often

Reply:
The full operation is done only when it is really required - if you remove dynamic rules, you must download them again (re-enabling), or when engines have been updated (once an hour of active usage of the extension).

Reply:
See this line, read comment
https://github.com/ghostery/ghostery-extension/pull/2593/files#diff-32fa3e38e24d85bd0f82578c7083f3c30d8642c1e65b42f5c7fab1511a17bbbeR53